kbmd/plugin.c

```diff
@@ -0,0 +1,1241 @@
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+#include <bunyan.h>
+#include <dirent.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <limits.h>
+#include <libnvpair.h>
+#include <libscf.h>
+#include <string.h>
+#include <strings.h>
+#include <umem.h>
+#include <unistd.h>
+#include <uuid/uuid.h>
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+#include <openssl/x509.h>
+#include <sys/sysmacros.h>
+#include "kbmd.h"
+#include "common.h"
+#include "ecustr.h"
+#include "envlist.h"
+#include "pivy/errf.h"
+#include "kspawn.h"
+#include "pivy/piv.h"
+#include "pivy/libssh/sshbuf.h"
+#include "pivy/libssh/ssherr.h"
+#include "pivy/libssh/sshkey.h"
+
+#include <stdio.h>
+
+/*
+ * In all cases, the PIV token must not currently be in a transaction
+ * when we call out to the plugins.  Since they are separate processes, they
+ * will be unable to use the PIV token if we keep it locked in a transaction.
+ */
+
+#define	PLUGIN_VERSION		"1"
+
+#define	PLUGIN_PREFIX		"kbm-plugin-"
+
+/* XXX: This path should change */
+#define	PLUGIN_PATH		"/usr/lib/kbm/plugins/"
+
+#define	GET_PIN_CMD		"get-pin"
+#define	REGISTER_TOK_CMD	"register-pivtoken"
+#define	REPLACE_TOK_CMD		"replace-pivtoken"
+#define	NEW_TOK_CMD		"new-rtoken"
+#define	POST_RECOVERY_UPDATE	"post-rcfg-update"
+
+extern char **_environ;
+
+static mutex_t plugin_mutex = ERRORCHECKMUTEX;
+static custr_t *kbmd_plugin;
+
+static size_t
+count_segments(const char *str, const char *sep)
+{
+	size_t n = 0;
+
+	while (*str != '\0') {
+		n++;
+
+		if ((str = strpbrk(str, sep)) == NULL)
+			return (n);
+		str++;
+	}
+
+	return (n);
+}
+
+/* Split src into multiple strings, separated be sep */
+static errf_t *
+split(custr_t *restrict src, const char *sep, custr_t ***restrict segsp,
+    size_t *restrict nsegp)
+{
+	errf_t *ret = ERRF_OK;
+	custr_t **segs = NULL;
+	size_t nsegs = 0;
+	const char *p = NULL;
+
+	nsegs = count_segments(custr_cstr(src), sep);
+
+	if ((ret = ecalloc(nsegs + 1, sizeof (custr_t *), &segs)) != ERRF_OK)
+		return  (ret);
+
+	p = custr_cstr(src);
+	for (size_t i = 0; i < nsegs; i++) {
+		custr_t *seg = NULL;
+		size_t len = 0;
+
+		if ((ret = ecustr_alloc(&seg)) != ERRF_OK)
+			goto done;
+
+		segs[i] = seg;
+
+		if ((len = strcspn(p, sep)) == 0)
+			len = strlen(p);
+
+		if ((ret = ecustr_append_printf(seg, "%.*s",
+		    (int)len, p)) != ERRF_OK)
+			goto done;
+
+		p += len + 1;
+	}
+
+	*segsp = segs;
+	*nsegp = nsegs;
+
+done:
+	if (ret != ERRF_OK) {
+		for (size_t i = 0; i < nsegs; i++)
+			custr_free(segs[i]);
+		free(segs);
+	}
+
+	return (ret);
+}
+
+/*
+ * Truncate the string cu to the first line, removing the trailing \n if
+ * present
+ */
+static void
+extract_line(custr_t *cu)
+{
+	const char *s = custr_cstr(cu);
+	const char *end = strchr(s, '\n');
+	size_t idx;
+
+	if (end == NULL)
+		return;
+
+	idx = (size_t)(end - s);
+	VERIFY0(custr_trunc(cu, idx));
```

Comment:
Does custr_trunc() have builtin support (from ecustr.c) for zeroing? Or does it not need it?

Reply:
Work I did a while ago allowed custr_ts to use a custom allocator.  All the custr_t's then use that allocator which does an explicit_bzero before freeing the memory (common/ecustr.c)
(the nvlist_t s also do a similar thing for similar reasons).

Reply:
Okay, the ecustr.c file changes that, excellent!